lib/rack/request.rb

```diff
@@ -480,25 +480,114 @@ def parseable_data?
         PARSEABLE_DATA_MEDIA_TYPES.include?(media_type)
       end
 
-      # Returns the data received in the query string.
-      def GET
-        if get_header(RACK_REQUEST_QUERY_STRING) == query_string
-          if query_hash = get_header(RACK_REQUEST_QUERY_HASH)
-            return query_hash
+      # Given a current input value, and a validity key, check if the cache
+      # is valid, and if so, return the cached value. If not, yield the
+      # current value to the block, and set the cache to the result.
+      #
+      # This method does not use cache_key, so it is shared between all
+      # instance of Rack::Request and it's sub-classes.
+      private def cache_for(key, validity_key, current_value)
+        # Get the current value of the validity key and compare it with the input value:
+        if get_header(validity_key).equal?(current_value)
+          # If the values are the same, then the cache is valid, so return the cached value.
+          if has_header?(key)
+            value = get_header(key)
+            # If the cached value is an exception, then re-raise it.
+            if value.is_a?(Exception)
+              raise value.class, value.message, cause: value.cause
+            else
+              # Otherwise, return the cached value.
+              return value
+            end
           end
         end
 
-        set_header(RACK_REQUEST_QUERY_HASH, expand_params(query_param_list))
+        # If the cache is not valid, then yield the current value to the block:
+        value = yield(current_value)
+
+        # Set the validity key to the current value so that we can detect changes:
+        set_header(validity_key, current_value)
+
+        # Set the cache to the result of the block, and return the result:
+        set_header(key, value)
+      rescue => error
+        # If an exception is raised, then set the cache to the exception, and re-raise it:
+        set_header(validity_key, current_value)
+        set_header(key, error)
+        raise
+      end
+
+      # This cache key is used by cached values generated by class_cache_for,
+      # specfically GET and POST. This is to ensure that the cache is not
+      # shared between instances of different classes which have different
+      # behaviour. This includes sub-classes that override query_parser or
+      # expand_params.
+      def cache_key
+        query_parser.class
```

Comment:
I still think this is a mistake, and think the onus should be on the framework to override this appropriately.  If you have to base it on the query parser, `query_parser` would be better than `query_parser.class`, considering that the default is to use a singleton `QueryParser` instance (`Utils.default_query_parser`), and singleton classes allow for the changing of behavior for individual instances.  It would be best to fix that first, and include that in the backport.